File: src/main/kotlin/cn/cangnova/cangjie/descriptors/impl/ClassConstructorDescriptorImpl.java
//package cn.cangnova.cangjie.descriptors.impl;
//
//import cn.cangnova.cangjie.descriptors.*;
//import cn.cangnova.cangjie.descriptors.annotations.Annotations;
//import cn.cangnova.cangjie.name.Name;
//import cn.cangnova.cangjie.name.SpecialNames;
//import cn.cangnova.cangjie.types.CangJieType;
//import cn.cangnova.cangjie.types.TypeSubstitutor;
//import org.jetbrains.annotations.NotNull;
//import org.jetbrains.annotations.Nullable;
//
//import java.util.Collection;
//import java.util.Collections;
//import java.util.List;
//
//public class ClassConstructorDescriptorImpl extends FunctionDescriptorImpl implements ClassConstructorDescriptor {
//
//    protected final boolean isPrimary;
//
//    protected ClassConstructorDescriptorImpl(
//            @NotNull ClassDescriptor containingDeclaration,
//            @Nullable ConstructorDescriptor original,
//            @NotNull Annotations annotations,
//            boolean isPrimary,
//            @NotNull Kind kind,
//            @NotNull SourceElement source
//
//    ) {
//
//        super(containingDeclaration, original, annotations, SpecialNames.INIT, kind, source);
//
//
//        this.isPrimary = isPrimary;
//    }
//
//
//
//    @NotNull
//    public static ClassConstructorDescriptorImpl create(
//            @NotNull ClassDescriptor containingDeclaration,
//            @NotNull Annotations annotations,
//            boolean isPrimary,
//            @NotNull SourceElement source
//
//    ) {
//        return create(containingDeclaration, annotations, isPrimary, source, false);
//    }
//
//    @NotNull
//    public static ClassConstructorDescriptorImpl create(
//            @NotNull ClassDescriptor containingDeclaration,
//            @NotNull Annotations annotations,
//            boolean isPrimary,
//            @NotNull SourceElement source,
//            boolean isEnd
//    ) {
//        return new ClassConstructorDescriptorImpl(containingDeclaration, null, annotations, isPrimary, Kind.DECLARATION, source);
//    }
//
//    @NotNull
//    public static ClassConstructorDescriptorImpl createSynthesized(
//            @NotNull ClassDescriptor containingDeclaration,
//            @NotNull Annotations annotations,
//            boolean isPrimary,
//            @NotNull SourceElement source
//    ) {
//        return new ClassConstructorDescriptorImpl(containingDeclaration, null, annotations, isPrimary, Kind.SYNTHESIZED, source);
//    }
//
//    public ClassConstructorDescriptorImpl initialize(
//            @NotNull List<ValueParameterDescriptor> unsubstitutedValueParameters,
//            @NotNull DescriptorVisibility visibility,
//            @NotNull List<TypeParameterDescriptor> typeParameterDescriptors
//    ) {
//        super.initialize(
//                null, calculateDispatchReceiverParameter(), calculateContextReceiverParameters(),
//                typeParameterDescriptors,
//                unsubstitutedValueParameters, null,
//                Modality.FINAL, visibility);
//        return this;
//    }
//
//    public ClassConstructorDescriptorImpl initialize(
//            @NotNull List<ValueParameterDescriptor> unsubstitutedValueParameters,
//            @NotNull DescriptorVisibility visibility
//    ) {
//        initialize(unsubstitutedValueParameters, visibility, getContainingDeclaration().getDeclaredTypeParameters());
//        return this;
//    }
//
//    @Nullable
//    public ReceiverParameterDescriptor calculateDispatchReceiverParameter() {
//        return null;
//    }
//
//    @NotNull
//    private List<ReceiverParameterDescriptor> calculateContextReceiverParameters() {
//        ClassDescriptor classDescriptor = getContainingDeclaration();
//        if (!classDescriptor.getContextReceivers().isEmpty()) {
//            return classDescriptor.getContextReceivers();
//        }
//        return Collections.emptyList();
//    }
//
//    @NotNull
//    @Override
//    public ClassDescriptor getContainingDeclaration() {
//        return (ClassDescriptor) super.getContainingDeclaration();
//    }
//
//    @NotNull
//    @Override
//    public ClassDescriptor getConstructedClass() {
//        return getContainingDeclaration();
//    }
//
//    @NotNull
//    @Override
//    public ClassConstructorDescriptor getOriginal() {
//        return (ClassConstructorDescriptor) super.getOriginal();
//    }
//
//    @Nullable
//    @Override
//    public ClassConstructorDescriptor substitute(@NotNull TypeSubstitutor originalSubstitutor) {
//        return (ClassConstructorDescriptor) super.substitute(originalSubstitutor);
//    }
//
//    @Override
//    public <R, D> R accept(DeclarationDescriptorVisitor<R, D> visitor, D data) {
//        return visitor.visitConstructorDescriptor(this, data);
//    }
//
//    @Override
//    public boolean isPrimary() {
//        return isPrimary;
//    }
//
//    @NotNull
//    @Override
//    public Collection<? extends FunctionDescriptor> getOverriddenDescriptors() {
//        return Collections.emptySet();
//    }
//
//    @Override
//    public void setOverriddenDescriptors(@NotNull Collection<? extends CallableMemberDescriptor> overriddenDescriptors) {
//        assert overriddenDescriptors.isEmpty() : "Constructors cannot override anything";
//    }
//
//    @NotNull
//    @Override
//    protected ClassConstructorDescriptorImpl createSubstitutedCopy(
//            @NotNull DeclarationDescriptor newOwner,
//            @Nullable FunctionDescriptor original,
//            @NotNull Kind kind,
//            @Nullable Name newName,
//            @NotNull Annotations annotations,
//            @NotNull SourceElement source
//    ) {
//        if (kind != Kind.DECLARATION && kind != Kind.SYNTHESIZED) {
//            throw new IllegalStateException("Attempt at creating a constructor that is not a declaration: \n" +
//                    "copy from: " + this + "\n" +
//                    "newOwner: " + newOwner + "\n" +
//                    "kind: " + kind);
//        }
//        assert newName == null : "Attempt to rename constructor: " + this;
//        return new ClassConstructorDescriptorImpl(
//                (ClassDescriptor) newOwner,
//                this,
//                annotations,
//                isPrimary,
//                Kind.DECLARATION,
//                source
//        );
//    }
//
//    @NotNull
//    @Override
//    public ClassConstructorDescriptor copy(
//            DeclarationDescriptor newOwner,
//            Modality modality,
//            DescriptorVisibility visibility,
//            Kind kind,
//            boolean copyOverrides
//    ) {
//        return (ClassConstructorDescriptor) super.copy(newOwner, modality, visibility, kind, copyOverrides);
//    }
//}

File: src/main/kotlin/com/linqingying/cangjie/descriptors/impl/ClassConstructorDescriptorImpl.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package com.linqingying.cangjie.descriptors.impl

import com.linqingying.cangjie.descriptors.*
import com.linqingying.cangjie.descriptors.annotations.Annotations
import com.linqingying.cangjie.name.Name
import com.linqingying.cangjie.name.SpecialNames
import com.linqingying.cangjie.types.CangJieType
import com.linqingying.cangjie.types.TypeSubstitutor


open class ClassConstructorDescriptorImpl protected constructor(
    containingDeclaration: ClassDescriptor,
    original: ConstructorDescriptor?,
    annotations: Annotations,
    private val isPrimary: Boolean,
    kind: CallableMemberDescriptor.Kind,
    source: SourceElement,
    private val isEnd: Boolean = false
) : FunctionDescriptorImpl(
    containingDeclaration, original, annotations, if (isEnd) {
        SpecialNames.END_INIT
    } else {
        SpecialNames.INIT
    }, kind, source
),
    ClassConstructorDescriptor {

    constructor(
        containingDeclaration: ClassDescriptor,
        original: ConstructorDescriptor?,
        annotations: Annotations,
        isPrimary: Boolean,
        kind: CallableMemberDescriptor.Kind,
        source: SourceElement,
    ) : this(containingDeclaration, original, annotations, isPrimary, kind, source, false)

    override fun getReturnType(): CangJieType {
        return super.getReturnType()!!
    }

    fun initialize(
        unsubstitutedValueParameters: List<ValueParameterDescriptor>,
        visibility: DescriptorVisibility,
        typeParameterDescriptors: List<TypeParameterDescriptor?>
    ): ClassConstructorDescriptorImpl {
        super.initialize(
            null, calculateDispatchReceiverParameter(), calculateContextReceiverParameters(),
            typeParameterDescriptors,
            unsubstitutedValueParameters, null,
            Modality.FINAL, visibility
        )
        return this
    }

    fun initialize(
        unsubstitutedValueParameters: List<ValueParameterDescriptor>
    ): ClassConstructorDescriptorImpl {
        initialize(
            unsubstitutedValueParameters,
            DescriptorVisibilities.PUBLIC,
            containingDeclaration.declaredTypeParameters
        )
        return this
    }

    fun initialize(
        unsubstitutedValueParameters: List<ValueParameterDescriptor>,
        visibility: DescriptorVisibility
    ): ClassConstructorDescriptorImpl {
        initialize(unsubstitutedValueParameters, visibility, containingDeclaration.declaredTypeParameters)
        return this
    }

    private fun calculateDispatchReceiverParameter(): ReceiverParameterDescriptor? {
        return null
    }

    private fun calculateContextReceiverParameters(): List<ReceiverParameterDescriptor> {
        val classDescriptor = containingDeclaration
        if (classDescriptor.contextReceivers.isNotEmpty()) {
            return classDescriptor.contextReceivers
        }
        return emptyList()
    }

    override val containingDeclaration: ClassDescriptor
        get() = super.containingDeclaration as ClassDescriptor

    override fun getConstructedClass(): ClassDescriptor {
        return containingDeclaration
    }

    override val original: ClassConstructorDescriptor
        get() = super.original as ClassConstructorDescriptor

    override fun substitute(substitutor: TypeSubstitutor): ClassConstructorDescriptor? {
        return super.substitute(substitutor) as ClassConstructorDescriptor?
    }

    override fun <R, D> accept(visitor: DeclarationDescriptorVisitor<R, D>, data: D?): R? {
        return visitor.visitConstructorDescriptor(this, data)
    }



    override fun isEnd(): Boolean {
        return isEnd
    }

    override fun isPrimary(): Boolean {
        return isPrimary
    }


    override fun getOverriddenDescriptors(): Collection<FunctionDescriptor> {
        return emptySet()
    }

    override fun setOverriddenDescriptors(overriddenDescriptors: Collection<CallableMemberDescriptor?>) {
        assert(overriddenDescriptors.isEmpty()) { "Constructors cannot override anything" }
    }

    override fun createSubstitutedCopy(
        newOwner: DeclarationDescriptor,
        original: FunctionDescriptor?,
        kind: CallableMemberDescriptor.Kind,
        newName: Name?,
        annotations: Annotations,
        source: SourceElement
    ): ClassConstructorDescriptorImpl {
        check(!(kind != CallableMemberDescriptor.Kind.DECLARATION && kind != CallableMemberDescriptor.Kind.SYNTHESIZED)) {
            """
                Attempt at creating a constructor that is not a declaration: 
                copy from: $this
                newOwner: $newOwner
                kind: $kind
                """.trimIndent()
        }
        assert(newName == null) { "Attempt to rename constructor: $this" }
        return ClassConstructorDescriptorImpl(
            newOwner as ClassDescriptor,
            this,
            annotations,
            isPrimary,
            CallableMemberDescriptor.Kind.DECLARATION,
            source
        )
    }

    override fun copy(
        newOwner: DeclarationDescriptor,
        modality: Modality,
        visibility: DescriptorVisibility,
        kind: CallableMemberDescriptor.Kind,
        copyOverrides: Boolean
    ): ClassConstructorDescriptor {
        return super.copy(newOwner, modality, visibility, kind, copyOverrides) as ClassConstructorDescriptor
    }

    companion object {
        @JvmOverloads
        fun create(
            containingDeclaration: ClassDescriptor,
            annotations: Annotations,
            isPrimary: Boolean,
            source: SourceElement,
            isEnd: Boolean = false
        ): ClassConstructorDescriptorImpl {
            return ClassConstructorDescriptorImpl(
                containingDeclaration,
                null,
                annotations,
                isPrimary,
                CallableMemberDescriptor.Kind.DECLARATION,
                source, isEnd
            )
        }

        fun createSynthesized(
            containingDeclaration: ClassDescriptor,
            annotations: Annotations,
            isPrimary: Boolean,
            source: SourceElement
        ): ClassConstructorDescriptorImpl {
            return ClassConstructorDescriptorImpl(
                containingDeclaration,
                null,
                annotations,
                isPrimary,
                CallableMemberDescriptor.Kind.SYNTHESIZED,
                source
            )
        }
    }
}
